Floquet changes

CamilleLCal opened the pull request:
**Description**
1) Changes in floquet_master_equation_rates to make it faster.
2) Correction transposition errors in floquet_master_equation_tensor and changes to make it faster. NB: This version works only for 1 rate matrix but could be generalized easily.
3) Correction in floquet_markov_mesolve. The basis change that returns the results in the computational had errors. NB: Results are still given in the interaction picture.

**Related issues or PRs**

**Changelog**
Fixed transposition error in floquet_master_equation_tensor
Fixed basis change error in floquet_markov_mesolve
Made floquet_master_equation_rates faster
Made floquet_master_equation_tensor faster

Ericgig commented:
I tried the notebook: https://github.com/qutip/qutip-notebooks/blob/master/examples/floquet-dynamics.ipynb
Under "Floquet-Markov master equation: comparison with other solvers", one of the result is strange with this version. 
I would change the Hamiltonian (everywhere is the notebook) from
`H = [H0, [H1, 'sin(w * t)']]`
to:
`H = [H0, [H1, lambda t, args: np.sin(args['w'] * t)]]`
as there is not reuse of the compiled string.

Ericgig commented on qutip/floquet.py:
```diff
@@ -707,28 +722,38 @@ def floquet_master_equation_tensor(Alist, f_energies):
+    ###
+    # Here there were transposition errors because the paper used in
+    # the documentation has typos.
+    # NB: Works only for 1 rate matrix.
```
Does that mean that Alist must have at most 1 element? 
Is it a new restriction? What happen if many rate matrix are given?

CamilleLCal replied:
I forgot to remove the comment "# NB: Works only for 1 rate matrix.", Alist can have as many elements as necessary.

CamilleLCal commented:
**Summary of what was done:**

- Changes in floquet_master_equation_rates to make it faster.
- Corrected transposition errors in _floquet_master_equation_tensor_ and changes to make it faster. I also removed a line so that the ME is solved in the interaction picture, this implies a different basis change in the _floquet_markov_mesolve_ function.
- Correction in _floquet_markov_mesolve_. As explained in the point above, the Floquet-Markov ME was not being solved in the interaction picture, therefore I modified _floquet_master_equation_tensor_ and modified the basis change in _floquet_markov_mesolve_. This modification requires an additional variable in the entry of _floquet_markov_mesolve_ which can not be chosen by default. For the moment, if this variable is not given and _floquet_basis=False_, then the density matrix or expectation values are  returned in the interaction picture, in the computational basis, and a warning message is sent. 
- Added integration options in: _floquet_modes, floquet_modes_t, floquet_modes_table, floquet_states_t, floquet_modes_t, floquet_wavefunction_t, fsesolve, floquet_master_equation_rates,  fmmesolve_
- Added test functions to verify the dissipative dynamics:
   a) Compare _fmmesolve_ and _mesolve_ for the cases: (NO drive but dissipation) and (Drive but NO dissipation)
   b) Compare numerical and analytical matrix elements for a driven-dissipative TLS with RWA

There are some places where the code could be modified to optimize the running time, especially in the basis change in _floquet_markov_mesolve_.

Ericgig commented:
@jakelishman Could you look at this? I beleive you have some experience with floquet equation and this should eventually go to `dev-major` so it would be nice if we could look at it before you finish the merge `master` -> `dev-major`.

jakelishman commented:
Sure, I'll try soon.  I can add stuff to the merge I've already done - I learned some `git` magic to combine merge commits into the same object without losing my previous work.

jakelishman commented on qutip/tests/test_floquet.py:
```diff
@@ -32,24 +32,267 @@
+        # Compare with results from standard schrodinger equation
+        sol_ref = mesolve(H, psi0, tlist, [], e_ops, args)
+
+        assert_(max(abs(sol.expect[0] - sol_ref.expect[0])) < 1e-4)
```
This sort of thing is better spelled as
```python
np.testing.assert_allclose(sol.expect[0], sol_ref.expect[0], atol=1e-4)
```
because then on a failure we'll see exactly what was in both arrays, and how much they differ by.

jakelishman commented on qutip/tests/test_floquet.py:
```diff
@@ -32,24 +32,267 @@
+            # Check energies
+            deltas = np.ndarray.flatten(DeltaMatrix)
+
+            assert_(min(abs(deltas-array_ana_delta[idx])) < 1e-4)
```
`min` doesn't seem right?  That succeeds as long as any pair of elements is close, rather than requiring _all_ pairs to be close.  If it _is_ meant to be the latter, it should again be `np.testing.assert_allclose`.

jakelishman commented on qutip/floquet.py:
```diff
@@ -176,7 +186,9 @@ def floquet_modes_t(f_modes_0, f_energies, t, H, T, args=None):
-def floquet_modes_table(f_modes_0, f_energies, tlist, H, T, args=None):
+
+def floquet_modes_table(f_modes_0, f_energies, tlist, H, T, args=None,
+                        options=Options()):
```
`Options()` is a mutable object, so shouldn't be used as a function default.  You can instead do
```python
def floquet_modes_table(..., options=None):
    options = options or Options()
    ...
```

jakelishman reviewed (changes requested):
Heya, thanks so much for looking at the Floquet module.  I haven't had time to properly properly dig into the transposition error to verify that (though I'm sure you knew what you were doing), but I also have a couple of other concerns that I've put in the reviews below.

jakelishman commented:
I couldn't work out how to attach a patch to a GitHub comment, so I've made PR (CamilleLCal/qutip#1) against your fork to show the sort of changes I think `floquet_markov_mesolve` could benefit from.

CamilleLCal commented:
Hi @jakelishman, thank you very much for looking at the Pull Request! I will reply to your questions/comments as soon as possible.

jakelishman commented:
Sounds good!  When I made the review, I actually didn't even realise how slow the module was at all - I didn't realise how much of an improvement your PR represented.  It'll still be good to possibly fix the quartic memory usage (I can help if necessary), but it's already an order of magnitude improvement on speed, not to mention the incorrect dynamics.

Ericgig commented on qutip/floquet.py:
```diff
@@ -716,130 +757,177 @@ def floquet_basis_transform(f_modes, f_energies, rho0):
-    elif isinstance(e_ops, list):
-
-        n_expt_op = len(e_ops)
+        store_states = options.store_states or False
```
```suggestion
        store_states = opt.store_states or False
```

Ericgig commented on qutip/floquet.py:
```diff
@@ -716,130 +757,177 @@ def floquet_basis_transform(f_modes, f_energies, rho0):
-
+                dtype = np.float64 if op.isherm else np.complex128
+                output.expect.append(np.zeros(len(tlist), dtype=dtype))
+        store_states = options.store_states or (n_expt_op == 0)
```
```suggestion
        store_states = opt.store_states or (n_expt_op == 0)
```

Ericgig commented:
@jakelishman I meet with @CamilleLCal and we went over your comments. Do you have time for another review. If you don't please tell us and we will do it.

There is a small change to be made in the qutip/qutip-notebooks#121 with `floquet_markov_mesolve`. @CamilleLCal said she will look at it in the following weeks.

jakelishman reviewed (commented):
I'm so sorry I ended up dropping the ball on this PR, particularly to Camille, since it was her work.  I just glanced through, and as best as I remember it, the major concerns I had are now solved (large duplication of transformation-handling code in the solver loop).  I think there's a more efficient method of constructing the sparse tensor operator in COO/CSR form, but this PR is already a huge improvement in speed on the previous one, and I very much don't want to hold it up any more.
